klippy/extras/bed_mesh.py

```diff
@@ -0,0 +1,617 @@
+    def __init__(self, config, gcode, fade_start, fade_end):
+        self.fade_start = fade_start
+        self.fade_end = fade_end
+        self.fade_dist = self.fade_end - self.fade_start
+        self.split_delta_z = config.getfloat(
+            'split_delta_z', .025, minval=0.01)
+        self.move_check_distance = config.getfloat(
+            'move_check_distance', 5., minval=3.)
+        self.z_mesh = None
+        self.gcode = gcode
+    def set_mesh(self, mesh):
+        self.z_mesh = mesh
+    def build_move(self, prev_pos, next_pos):
+        self.prev_pos = tuple(prev_pos)
+        self.next_pos = tuple(next_pos)
+        self.current_pos = list(prev_pos)
+        gc_z = next_pos[2] - self.gcode.base_position[2]
+        self.z_factor = 1.
+        if gc_z > self.fade_start and self.fade_end > self.fade_start:
+            self.z_factor = (self.fade_end - gc_z) / self.fade_dist
+        self.z_offset =  \
+            self.z_factor \
+            * self.z_mesh.get_z(
+                self.prev_pos[0] - self.gcode.base_position[0],
+                self.prev_pos[1] - self.gcode.base_position[1])
+        self.traverse_complete = False
+        self.distance_checked = 0.
+        axes_d = [self.next_pos[i] - self.prev_pos[i] for i in range(4)]
+        self.total_move_length = math.sqrt(sum([d*d for d in axes_d[:3]]))
+        self.axis_move = [not isclose(d, 0., abs_tol=1e-10) for d in axes_d]
+    def _set_next_move(self, distance_from_prev):
+        t = distance_from_prev / self.total_move_length
+        if t > 1. or t < 0.:
+            self.gcode.respond_error(
+                "bed_mesh: Slice distance is negative "
+                "or greater than entire move length")
+            self.current_pos[:] = self.next_pos
+            return
+        for i in range(4):
+            if self.axis_move[i]:
+                self.current_pos[i] = lerp(
+                    t, self.prev_pos[i], self.next_pos[i])
+    def split(self):
+        if not self.traverse_complete:
+            if self.axis_move[0] or self.axis_move[1]:
+                # X and/or Y axis move, traverse if necessary
+                while self.distance_checked + self.move_check_distance \
+                        < self.total_move_length:
+                    self.distance_checked += self.move_check_distance
+                    self._set_next_move(self.distance_checked)
+                    next_z = \
+                        self.z_factor \
+                        * self.z_mesh.get_z(
+                            self.current_pos[0] - self.gcode.base_position[0],
+                            self.current_pos[1] - self.gcode.base_position[1])
+                    if abs(next_z - self.z_offset) >= self.split_delta_z:
+                        self.z_offset = next_z
+                        return self.current_pos[0], self.current_pos[1], \
+                            self.current_pos[2] + self.z_offset, \
+                            self.current_pos[3]
```

Comment:
Interesting.  I would have done this a different way - I would have implemented the mesh as a triangular mesh, and then always done linear interpolation between mesh points and always do line splitting at the point a move crosses a triangle border.  The "triangular mesh" (a series of equal sized triangles evenly spaced throughout the bed) could be filled directly by probing each triangle point or by using the bicubic/lagrange algos on a coarsely probed grid.

I suspect the method used here of splitting moves heuristically may result in some quirks.  For example, if a user does a printing move from 0,0 => 10,0 and then later does a non printing move from -5,5 => 5,5 then the toolhead may not be at the same Z position when it crosses the line at position 0,5.  (Using linear interpolation within a triangular mesh wouldn't have this quirk.)

That said, I don't know how other firmwares implement mesh bed leveling, and I don't know if the triangular method would have other quirks.  So, this is just a high level comment.